### DISCUSSION TITLE
Ability to time out on call to Exporter::Export

### DISCUSSION THREAD

**Discussion by ryan-tarr-ai** [assimp/assimp#5452] [Q&A]
Is there a way to have the function time out on a call to `Exporter::Export`? If not is there a best practice workaround?

I have a pipeline that converts obj to glb2 using `Importer::ReadFile` and then `Exporter::Export`. I also have a step in the middle to run `Importer::ApplyPostProcessing` with the `aiProcess_ValidateDataStructure` flag. This works fine for almost all cases, but recently ran into a case where the `Export` step would take ~70 hours (for an aiMesh with ~4 mil vertices) because the underlying `aiMesh` was malformed (due to a single vertex in the imported obj being `NaN`) which led to a `SpatialSort` step (`SpatialSort:: FindIdenticalPositions`) overcomputing (comparator with `NaN` in a while loop led to an always false case, basically causing a hang bc made this loop O(N^2) which is bad for 4 mil vertices) and taking forever to complete.

I was able to fix by adding the pFlag `aiProcess_FindInvalidData` to the `ApplyPostProcessing` step and see that the post process step now returns a nullptr (so I can now just raise an exception before I even get to the `Export` step). However, ideally there is a way to time out the Export call to prevent such a case from happening again. It is so much worse to hang for many hours than to error/time out.

**Comment by turol** [comment]
Not really. There's no reasonable way to cancel a function that wasn't explicitly built to support it or have support from a language runtime (which C++ doesn't have). You can try `pthread_cancel` but there's a high chance it will leave your program in an inconsistent state and definitely will leak memory.

The best way is to run each model import/export in its own process and then kill it if it's taking too long.

**Comment by RichardTea** [comment]
NaN should never occur for mesh data, the OBJ importer should have rejected that model (or at least the affected mesh) during the import.

Can you write up a bug for that and attach a sample model?
